docs: add a section about custom authentication supplier (#80)
neo4j/docs-spark: modules/ROOT/pages/configuration.adoc

        .option("labels", "Product")
        .load()
----
== Custom authentication supplier label:new[v.5.3.9]

> Starting from version 5.3.9 Neo4j Spark Connectors provides an option to use custom authentication supplier in addition to

Reviewer: ```suggestion
Starting from version 5.3.9, the connector provides an option to use a custom authentication supplier in addition to
```

already supported authentication types, such as `none`, `basic`, `kerberos`, `custom` and `bearer`.

> Authentication supplier needs to implement interface `org.neo4j.connectors.authn.AuthenticationTokenSupplierFactory` from `org.neo4j.connectors:commons-authn-spi` library.

Reviewer: ```suggestion
The authentication supplier needs to implement `org.neo4j.connectors.authn.AuthenticationTokenSupplierFactory` from the `org.neo4j.connectors:commons-authn-spi` library.
```

The interface requires implementing two methods:

* *String getName()* - returns the name of the supplier.
Must be unique and not clash with existing authentication types.

* *Supplier<AuthenticationToken> create(String, String, Map<String, String>)* - accepts username, password and additional parameters and return a supplier of `AuthenticationToken`.

> An `AuthenticationToken` instances is an abstract representation of an arbitrary authentication to be presented to the Neo4j server and can contain a principal and its credentials (username and password), a token or other ways of confirming an identity.

Reviewer: ```suggestion
An `AuthenticationToken` instance is an abstract representation of an arbitrary authentication to be presented to the Neo4j server and can contain a principal and its credentials (username and password), a token or other ways of confirming an identity.
```

The interface itself offers the following factories:

* `AuthenticationToken#bearer`
* `AuthenticationToken#kerberos`
* `AuthenticationToken#none`
* `AuthenticationToken#usernameAndPassword`
* `AuthenticationToken#custom`

The typical use-case for an authentication supplier is supporting expiring tokens issued by OAuth 2.0 or OIDC providers.

Custom authentication supplier can be used via the service loader mechanism and setting configuration parameter `authentication.type` to the name
of the supplier. Additional parameters can be as configuration options prefixed with `authentication.$name.`.

=== Keycloak authentication supplier example

Keycloak authentication supplier in implemented in the `org.neo4j.connectors:commons-authn-keycloak` library.
It consists of a factory class creating an instance of an authentication token supplier;
[source, java]
----
include::example$java/KeycloakOIDCAuthenticationSupplierFactory.java[]
----

and a supplier that collects an access token from a Keycloak server and refreshes it on expiration.

[source, java]
----
include::example$java/KeycloakOIDCAuthenticationSupplier.java[]
----

Configuration options for it would look like

> [source, scala]

Reviewer 2: Why is this example in Scala while the rest is for Java?

An `AuthenticationToken` instance is an abstract representation of an arbitrary authentication to be presented to the Neo4j server and can contain a principal and its credentials (username and password), a token or other ways of confirming an identity.

The interface itself offers the following factories:

* `AuthenticationToken#bearer`
* `AuthenticationToken#kerberos`
* `AuthenticationToken#none`
* `AuthenticationToken#usernameAndPassword`
* `AuthenticationToken#custom`

The typical use-case for an authentication supplier is supporting expiring tokens issued by OAuth 2.0 or OIDC providers.

Custom authentication supplier can be used via the service loader mechanism and by setting configuration parameter `authentication.type` to the name
of the supplier. Additional parameters can be specified as configuration options prefixed with `authentication.$name.`.

=== Keycloak authentication supplier example

An authentication supplier for link:https://www.keycloak.org/[Keycloak] is provided as part of the `org.neo4j.connectors:commons-authn-keycloak` library.
It includes:

> * A factory class creating an instance of an authentication token supplier:
> [source, java]

Reviewer 2: ```suggestion
* A factory class creating an instance of an authentication token supplier:
+
[source, java]
```

----
include::example$java/KeycloakOIDCAuthenticationSupplierFactory.java[]
----

* A supplier that collects an access token from a Keycloak server and refreshes it on expiration:

> 

Reviewer 2: ```suggestion
+
```
